core/dbt/include/global_project/macros/materializations/helpers.sql

```diff
@@ -72,3 +72,96 @@
   {% endif %}
   {% do return(config_full_refresh) %}
 {% endmacro %}
+
+
+{% macro incremental_validate_on_schema_change(on_schema_change, default_value='ignore') %}
+   
+   {% if on_schema_change not in ['sync', 'append', 'fail', 'ignore'] %}
```

Comment:
- Whoa, `sync` is very cool! Could we make this even more explicit: `sync_all_columns`?
- I hesitate with the name `append`, since that (to my mind) always means adding _rows_, rather than adding _columns_. I get what you mean—add columns without removing any—but I wonder if we could find a different word here (or be more verbose: `add_new_columns`).
- The original proposal in #1132 also had a `refresh` option, motivated by the cases where users would want to fully backfill new columns (rather than null values for old rows). What do you think? Better to `'fail'` and give the user an explicit error telling them to manually refresh? I could go either way here.

Reply:
@jtcohen6 On board with those naming changes (`sync` --> `sync_all_columns` and `append` --> `add_new_columns`). In general I like a bit of verbosity up to a point :). 

On the third point about refresh behavior, I think that should be handled in a separate unit of work, and that this current approach should update the columns and notify the user of what happened. I have a start on some additional code that would backfill columns according to data type (e.g. an INT column gets -1 for everything), with those default values specified in a mapping input to the materialization config.. Are you thinking we'd (optionally) backfill with actual data as well?

Reply:
For simplicity's sake, I think the default behavior for syncing/adding columns should be the database's default behavior, i.e. backfilling with null values. (I'm hesitant about having users supply a full mapping of every column and its default value in the case of schema evolution, though we could consider adding `default` as a property in the model's `columns` YAML dict.)

I do think, if a user wants the _actual_ values, that should require a full refresh. The question is, should full refreshing the table, in the case that a schema change is detected, be an optional strategy of `on_schema_change`? Or must a user manually pass the `--full-refresh` flag / set the `should_full_refresh` config?

Reply:
I could see use cases where making that automated full refresh configurable would be valuable. That would actually be relatively straightforward to program with a config like you described.

Reply:
So, NULLs when new columns are added is the new default behavior, and optionally users can trigger a full refresh when either `append_new_columns` or `sync_all_columns` is specified.

Reply:
Hm, I think what I'm sayig is, the `full_refresh` option is orthogonal to sync/add columns, and it also lets us keep the sync/add options simpler.

By making `full_refresh` another potential value of `on_schema_change`, we're letting the user pick any of:
- Just add new columns (backfilled with nulls)
- Sync new + existing columns (backfilled with nulls)
- Rebuild the entire table from scratch (real backfill, but also $$$)
- Ignore (current behavior)
- Aggressively raise an error

The status quo is, the user needs to manually determine if there's a schema change, and if so, pass `--full-refresh`. This would be an improvement insofar as the incremental model would only full refresh if dbt determines it needs to.

Reply:
Ooooh! So we'd have the possible values of `on_schema_change` be `['fail', 'ignore', 'sync_all_columns', 'append_new_columns', 'full_refresh']` then?

Reply:
That's what I'm thinking... what say you?

Reply:
Makes sense! In my brain the "full refresh in the case of schema change" functionality was a *separate* config set as a True/False, so comes to the same thing behavior wise, but this is a cleaner way to expose it to the user.